README.md
<!-- These are examples of badges you might also want to add to your README. Update the URLs accordingly.
[![Project generated with PyScaffold](https://img.shields.io/badge/-PyScaffold-005CA0?logo=pyscaffold)](https://pyscaffold.org/)
[![Built Status](https://api.cirrus-ci.com/github/<USER>/classify_covid.svg?branch=main)](https://cirrus-ci.com/github/<USER>/classify_covid)
[![ReadTheDocs](https://readthedocs.org/projects/classify_covid/badge/?version=latest)](https://classify_covid.readthedocs.io/en/stable/)
[![Coveralls](https://img.shields.io/coveralls/github/<USER>/classify_covid/main.svg)](https://coveralls.io/r/<USER>/classify_covid)
[![PyPI-Server](https://img.shields.io/pypi/v/classify_covid.svg)](https://pypi.org/project/classify_covid/)
[![Conda-Forge](https://img.shields.io/conda/vn/conda-forge/classify_covid.svg)](https://anaconda.org/conda-forge/classify_covid)
[![Monthly Downloads](https://pepy.tech/badge/classify_covid/month)](https://pepy.tech/project/classify_covid)
[![Twitter](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Twitter)](https://twitter.com/classify_covid)
-->

# Covid Classification for Mobile Inference

> This repository contains code for modelling, integrations, tests, and documentation for covid classification

## VSCode Setup

In order to set up the necessary container:

1. Setup Github with SSH and Install Docker Desktop like mentioned in ML-Team-Guidelines
2. clone this repository and open the repo folder insider vscode
   ```bash
   keyboard command : left CTRL + Shift + P
   Type: ">Dev Containers: Rebuild Container"
   ```

## Installation

In order to set up the necessary environment:

1. review and uncomment what you need in `environment.yml` and create an environment `classify_covid` with the help of [conda]:
   ```bash
   conda env create -f requirements/<framework>/environment.yml
   e.g. conda env create -f requirements/tensorflow/environment.yml
   ```
2. activate the new environment with:
   ```bash
   conda activate <framework>
   e.g. conda activate tensorflow
   ```
   ```bash
   python -m ipykernel install --user --name=<framework>
   e.g. python -m ipykernel install --user --name=tensorflow
   ```

> **_NOTE:_**  The conda environment will have classify_covid installed in editable mode.
> Some changes, e.g. in `setup.cfg`, might require you to run `pip install -e .` again.


Optional and needed only once after `git clone`:

3. install several [pre-commit] git hooks with:
   ```bash
   pre-commit install
   # You might also want to run `pre-commit autoupdate`
   ```
   and checkout the configuration under `.pre-commit-config.yaml`.
   The `-n, --no-verify` flag of `git commit` can be used to deactivate pre-commit hooks temporarily.

4. to run tests manually you can use below function

   ```bash
   pre-commit run --all-files
   # we suggest to use this regularly in order to avoid errors
   ```

5. to start jupyterlab, you can use below command
   ```bash
   jupyter lab --no-browser --port=8890 --allow-root --ip='0.0.0.0' --NotebookApp.token='' --NotebookApp.password=''
   # In order to open jupyterlab, you can go to ports > globe icon beside local address link
   ```

<!--
4. install [nbstripout] git hooks to remove the output cells of committed notebooks with:
   ```bash
   nbstripout --install --attributes notebooks/.gitattributes
   ```
   This is useful to avoid large diffs due to plots in your notebooks.
   A simple `nbstripout --uninstall` will revert these changes.

Then take a look into the `scripts` and `notebooks` folders.

## Dependency Management & Reproducibility

1. Always keep your abstract (unpinned) dependencies updated in `environment.yml` and eventually
   in `setup.cfg` if you want to ship and install your package via `pip` later on.
2. Create concrete dependencies as `environment.lock.yml` for the exact reproduction of your
   environment with:
   ```bash
   conda env export -n classify_covid -f environment.lock.yml
   ```
   For multi-OS development, consider using `--no-builds` during the export.
3. Update your current environment with respect to a new `environment.lock.yml` using:
   ```bash
   conda env update -f environment.lock.yml --prune
   ```
-->

## Project Organization

```
├── AUTHORS.md              <- List of developers and maintainers.
├── CHANGELOG.md            <- Changelog to keep track of new features and fixes.
├── CONTRIBUTING.md         <- Guidelines for contributing to this project.
├── Dockerfile              <- Build a docker container with `docker build .`.
├── LICENSE.txt             <- License as chosen on the command-line.
├── README.md               <- The top-level README for developers.
├── configs                 <- Directory for configurations of model & application.
├── requirements
│   ├── tensorflow          <- contains tensorflow environment and requirements
│   │   └── environment.yml 
│   └── pytorch             <- contains pytorch environment and requirements
│       └── environment.yml 
├── data
│   ├── external            <- Data from third party sources.
│   ├── interim             <- Intermediate data that has been transformed.
│   ├── processed           <- The final, canonical data sets for modeling.
│   └── raw                 <- The original, immutable data dump.
│
├── docs                    <- Directory for Sphinx documentation in rst or md.
├── environment.yml         <- The conda environment file for reproducibility.
├── models                  <- Trained and serialized models, model predictions,
│                              or model summaries.
├── notebooks               <- Jupyter notebooks. Naming convention is a number (for
│                              ordering), the creator's initials and a description,
│                              e.g. `1.0-fw-initial-data-exploration`.
├── pyproject.toml          <- Build configuration. Don't change! Use `pip install -e .`
│                              to install for development or to build `tox -e build`.
├── references              <- Data dictionaries, manuals, and all other materials.
├── reports                 <- Generated analysis as HTML, PDF, LaTeX, etc.
│   └── figures             <- Generated plots and figures for reports.
├── scripts                 <- Analysis and production scripts which import the
│                              actual PYTHON_PKG, e.g. train_model.
├── setup.cfg               <- Declarative configuration of your project.
├── setup.py                <- [DEPRECATED] Use `python setup.py develop` to install for
│                              development or `python setup.py bdist_wheel` to build.
├── src
│   └── classify_covid      <- Actual Python package where the main functionality goes.
├── tests                   <- Unit tests which can be run with `pytest`.
├── .coveragerc             <- Configuration for coverage reports of unit tests.
├── .isort.cfg              <- Configuration for git hook that sorts imports.
└── .pre-commit-config.yaml <- Configuration of pre-commit git hooks.
```

<!-- pyscaffold-notes -->

## Note

This project has been set up using [PyScaffold] 4.3.1 and the [dsproject extension] 0.7.2.

[conda]: https://docs.conda.io/
[pre-commit]: https://pre-commit.com/
[Jupyter]: https://jupyter.org/
[nbstripout]: https://github.com/kynan/nbstripout
[Google style]: http://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings
[PyScaffold]: https://pyscaffold.org/
[dsproject extension]: https://github.com/pyscaffold/pyscaffoldext-dsproject


CHANGELOG.md
# Changelog

## Version 0.1 (development)

1. Added pip dependencies to tensorflow env file - dvc[all], wandb, hydra-core, scikit-learn, __pillow__(Necessary for this model. not installing pillow throws error.)

2. `wandb login`

3. `dvc init`

4. Loaded training data to _data/raw_ folder.

5. Remove "" from config file.

5. Removed _data_ from _git_. 

```bash
git rm -r --cached 'data'
```

> Trying to add to DVC versioning takes long time on docker container on local machine. Suggested to do on server only.

6. In file __train_model.py__
    1. Added configuration to WandB - `wandb.init(project="classify_covid", config=cfg)`
    2. Added main function with arguments parser

    ```python
    
    if __name__ == "__main__":
        argparser = argparse.ArgumentParser()
        argparser.add_argument('--config', dest='config', required=True)
        args = argparser.parse_args()

        train(config_name=args.config)
    ```

    3. Converted the file structure to a function `train(config_name: Text) -> None:`

    4. Changed `conf/` to `configs/`

    5. Changed loading of _c, n, p_ from direct paths to path from config file. 

    6. Added `model.save()` to save model post training.

    7. Changed `lr = 0.01` to `lr = cfg.params.learning_rate` 

7. Added model save location to _config.yaml_

8. Added stage in dvc for training model.

```bash
dvc stage add -n train -p configs/config.yaml -o models/covid.h5 scripts/train_model.py --config=config.yaml
# not added -d data/raw, since data is not cached. For training in server add: -d data/raw
```
9. Added `model.py` & `preprocessing.py` to _dvc.yaml: deps_

10. Added `test_model.py` in `scripts` to predict model outcome on new images.

11. Added tests in `tests` folder.


tests/conftest.py
"""
    Dummy conftest.py for classify_covid.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    - https://docs.pytest.org/en/stable/fixture.html
    - https://docs.pytest.org/en/stable/writing_plugins.html
"""

# import pytest


scripts/model.py
import os

import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow.keras.applications.densenet import DenseNet121
from tensorflow.keras.layers import AveragePooling2D, Dense, Dropout, Flatten
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import *


def get_Model(freeze):  #model_cfg as input
  # cfg = model_cfg

  baseModel = DenseNet121(input_shape=(224, 224, 3), weights='imagenet', include_top=False)

  if freeze:
    for layer in baseModel.layers:
      layer.trainable = False

  headModel = baseModel.output
  headModel = AveragePooling2D()(headModel)
  headModel = Flatten()(headModel)
  headModel = Dense(128, activation="relu")(headModel)
  headModel = Dropout(0.2)(headModel)
  headModel = Dense(3, activation='softmax')(headModel)

  model = Model(inputs=baseModel.input, outputs=headModel)
  return model

scripts/preprocessing.py
import os

import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator


def create_df(filenames, c, p, n):
  categories = []
  for filename in filenames:
    category = filename.split('-')[0]
    if category == 'COVID':
      categories.append(str(2))
    elif category == 'Viral Pneumonia':
      categories.append(str(1))
    else:
      categories.append(str(0))

  """converting to dataframe"""
  for i in range(len(filenames)):
    if 'COVID' in filenames[i]:
      filenames[i] = os.path.join(c, filenames[i])
    elif 'Viral Pneumonia' in filenames[i]:
      filenames[i] = os.path.join(p, filenames[i])
    else:
      filenames[i] = os.path.join(n, filenames[i])

  df = pd.DataFrame({
    'filename': filenames,
    'category': categories
  })
  return df

def train_generator(train_data):
    train_data_gen = ImageDataGenerator(
    rotation_range=15,
    rescale=1./255,
    shear_range=0.1,
    zoom_range=0.2,
    horizontal_flip=True,
    width_shift_range=0.1,
    height_shift_range=0.1
    )
    train_generator = train_data_gen.flow_from_dataframe(
        train_data,
        x_col='filename',
        y_col='category',
        target_size=(224,224),
        class_mode='categorical',
        batch_size=15
    )
    return train_generator

def val_test_generator(data):
    val_test_data_gen = ImageDataGenerator(rescale=1./255)

    val_test_generator = val_test_data_gen.flow_from_dataframe(
        data,
        x_col='filename',
        y_col='category',
        target_size=(224,224),
        class_mode='categorical',
        batch_size=15
    )
    return val_test_generator


scripts/train_model.py
import argparse
import os
import random
from typing import Text

import numpy as np
import pandas as pd
import tensorflow as tf
import wandb.keras
from hydra import compose, initialize
from model import get_Model
from omegaconf import OmegaConf
from PIL import Image
from preprocessing import create_df, train_generator, val_test_generator
from sklearn.model_selection import train_test_split
from tensorflow.keras.optimizers import Adam

import wandb


def train(config_name: Text) -> None:
    # set the random seeds

    # os.environ['TF_CUDNN_DETERMINISTIC'] = '2'
    random.seed(hash("setting random seeds") % 2**32 - 1)   #to control randomness
    np.random.seed(hash("improves reproductibility") % 2**32 - 1) #to make it reproducable as possible
    tf.random.set_seed(hash("by removing stochasticity") % 2**32 - 1)


    initialize(version_base=None, config_path="../configs")
    cfg=compose(config_name=config_name)
    print(OmegaConf.to_yaml(cfg))

    wandb.init(project="classify_covid", config=cfg)

    #data
    c = cfg.data.covid
    n = cfg.data.normal
    p = cfg.data.viral

    path='/workspaces/classify_covid/data/raw/df_csv.csv'

    if not os.path.exists(path): create_df(c, p, n)

    df= pd.read_csv(path, sep=',',index_col=False)
    df = df.astype(str)
    # print(df)
    # print(df.dtypes)

    #train_test_split
    train_data, test_valid_data = train_test_split(df, test_size=0.2, random_state = 42, shuffle=True, stratify=df['category'])
    train_data = train_data.reset_index(drop=True)

    test_valid_data = test_valid_data.reset_index(drop=True)

    test_data, valid_data = train_test_split(test_valid_data, test_size=0.5, random_state = 42,
                                             shuffle=True, stratify=test_valid_data['category'])
    test_data = test_data.reset_index(drop=True)
    valid_data = valid_data.reset_index(drop=True)

    #data_augmentation
    train_gen = train_generator(train_data)
    valid_gen = val_test_generator(valid_data)
    test_gen = val_test_generator(test_data)

    model = get_Model(freeze=True)
    # loss=cfg.params.loss
    #lr=0.01

    lr = cfg.params.learning_rate
    epoch_num=cfg.params.training_epoch
    model.compile(optimizer=Adam(learning_rate=lr), metrics=["accuracy"], loss='categorical_crossentropy')

    history = model.fit(train_gen,
                        validation_data=valid_gen,
                        verbose=1,
                        epochs=epoch_num,
                        callbacks=[wandb.keras.WandbCallback()])
    
    model.save(cfg.model.save)

if __name__ == "__main__":
    argparser = argparse.ArgumentParser()
    argparser.add_argument('--config', dest='config', required=True)
    args = argparser.parse_args()

    train(config_name=args.config)